router/http.go

```diff
@@ -422,109 +383,271 @@ func (s *httpService) getBackendSticky(req *http.Request) (*httputil.ClientConn,
-			clientIP = strings.Join(prior, ", ") + ", " + clientIP
+	isSticky := false
+	stickyAddr := ""
+	if req.Header.Get(hdrUseStickySessions) == "true" {
+		// TODO(bgentry): switch to better way to check sticky setting
+		isSticky = true
+		req.Header.Del(hdrUseStickySessions)
+		if stickyAddr = s.stickyCookieAddr(req); stickyAddr != "" {
+			sortStringFirst(addrs, stickyAddr)
 		}
-		req.Header.Set("X-Forwarded-For", clientIP)
 	}
-	if tls {
-		req.Header.Set("X-Forwarded-Proto", "https")
-	} else {
-		req.Header.Set("X-Forwarded-Proto", "http")
-	}
-	// TODO: Set X-Forwarded-Port
+
+	// Most of this is borrowed from httputil.ReverseProxy
+	outreq := new(http.Request)
+	*outreq = *req // includes shallow copies of maps, but okay
 
 	// Pass the Request-URI verbatim without any modifications
-	req.URL.Opaque = strings.Split(strings.TrimPrefix(req.RequestURI, req.URL.Scheme+":"), "?")[0]
+	outreq.URL.Opaque = strings.Split(strings.TrimPrefix(req.RequestURI, req.URL.Scheme+":"), "?")[0]
+	outreq.URL.Scheme = "http"
+	outreq.Proto = "HTTP/1.1"
+	outreq.ProtoMajor = 1
+	outreq.ProtoMinor = 1
+	outreq.Close = false
 
-	if err := backend.Write(req); err != nil {
-		log.Println("server write err:", err)
-		// TODO: return error to client here
-		return true
+	// TODO: Proxy HTTP CONNECT? (example: Go RPC over HTTP)
+
+	// Directly bridge `Connection: Upgrade` requests
+	if strings.ToLower(outreq.Header.Get("Connection")) == "upgrade" {
+		s.forwardAndProxyTCP(w, outreq, addrs, stickyAddr, isSticky)
+		return
 	}
-	res, err := backend.Read(req)
-	if res != nil {
-		if stickyCookie != nil {
-			res.Header.Add("Set-Cookie", stickyCookie.String())
-		}
-		if res.StatusCode == http.StatusSwitchingProtocols {
-			res.Body = nil
+
+	// Remove hop-by-hop headers to the backend.  Especially
+	// important is "Connection" because we want a persistent
+	// connection, regardless of what the client sent to us.  This
+	// is modifying the same underlying map from req (shallow
+	// copied above) so we only copy it if necessary.
+	copiedHeaders := false
+	for _, h := range hopHeaders {
+		if outreq.Header.Get(h) != "" {
+			if !copiedHeaders {
+				outreq.Header = make(http.Header)
+				copyHeader(outreq.Header, req.Header)
+				copiedHeaders = true
+			}
+			outreq.Header.Del(h)
 		}
-		if err := sc.Write(req, res); err != nil {
-			if err != io.EOF && err != httputil.ErrPersistEOF {
-				log.Println("client write err:", err)
-				// TODO: log error
+	}
+
+	var (
+		backend string
+		res     *http.Response
+		err     error
+	)
+
+	for _, backend = range addrs {
+		// TODO: limit number of backends tried
+		// TODO: temporarily quarantine failing backends
+
+		outreq.URL.Host = backend
+		res, err = transport.RoundTrip(outreq)
+		if err != nil {
+			if _, ok := err.(*dialErr); ok {
+				// retry, maybe log a message about it
+				continue
 			}
-			return true
+			log.Println("http: proxy error:", err)
+			fail(w, 503, http.StatusText(503))
```

Comment:
yeah, as long as you're ok with always using the default text for that status code. Looks like that's pretty much what's already happening in the other places that were using it.